File: 프로그래머스/2/77885. 2개 이하로 다른 비트/2개 이하로 다른 비트.kt
class Solution {
    fun solution(numbers: LongArray): LongArray {
        val answer = LongArray(numbers.size)

        for (i in numbers.indices) {
            val x = numbers[i]
            
            answer[i] = if (x % 2L == 0L) {
                // 짝수면 +1
                x + 1
            } else {
                // 홀수면 오른쪽 첫 0 비트 찾기
                var bit = 1L
                while ((x and bit) != 0L) {
                    bit = bit shl 1
                }
                // 찾은 0 위치 바로 오른쪽 1비트만 더함
                x + (bit shr 1)
            }
        }

        return answer
    }
}

File: 프로그래머스/2/77885. 2개 이하로 다른 비트/2개 이하로 다른 비트.js
function solution(numbers) {
    var answer = [];
    
    for (let x of numbers) {
        if (x % 2 === 0) {
            // 짝수면 +1
            answer.push(x + 1);
        } else {
            // 홀수면 오른쪽 첫 0 비트 찾기
            let bit = 1n; // BigInt 사용 (x가 10^15까지 가능)
            let bigX = BigInt(x);
            
            while ((bigX & bit) !== 0n) {
                bit <<= 1n;
            }
            
            // 찾은 0 위치 바로 오른쪽 1비트만 더함
            answer.push(Number(bigX + (bit >> 1n)));
        }
    }
    
    return answer;
}
